### ISSUE TITLE
Suggested change to the miss chance calculation

### ISSUE THREAD

**Issue by ohganot** [LightsHope/server#755]
**Current formula**
```
    if (GetTypeId() == TYPEID_PLAYER && pVictim->GetTypeId() == TYPEID_PLAYER)
        hitChance = 95.0f + skillDiff * 0.04f;  // PvP misschance base is 5.00%
    else if (pVictim->GetTypeId() == TYPEID_PLAYER)
        hitChance = 94.4f + skillDiff * 0.04f;
    else if (skillDiff < -10)
        hitChance = 93.4f + (skillDiff + 10) * 0.4f;        // 7% ~ 6.60% base chance to miss for big skill diff
    else
        hitChance = 94.4f + skillDiff * 0.1f;
```

This results in the following miss chance values for weapon skill values between 300 and 315 vs a lvl 63 mob:
https://gyazo.com/3c7704a97c477b41e429146f9162dd35

**Issue**
First, lets look at the biggest issue in this formula. The base miss chance changes depending on the victim mob level. The base miss chance should always stay the same, modified by some sort of formula.

The base chances for dodge, parry, block are all 5%, increased by 0.04% for the each point of difference between player weapon skill and mob defense skill.
Source: http://wowwiki.wikia.com/wiki/Attack_table?oldid=338686

We know that miss chance doesn't follow the same rule, since it was a known fact back in the day that hit cap for yellow attacks is 8%.
Source: http://www.lurkerlounge.com/forums/thread-4078.html

The value of 5.6% comes from the fact that dodge, miss, chance, parry all have a base chance of 5%, increased by the difference in player weapon skill and mob defense skill.

Therefore when a player has 315 weapon skill and a mob has 315 defense skill, the differences should equal out resulting in 5% miss chance, which is not currently the case.

**Suggested formula**
```
    if (GetTypeId() == TYPEID_PLAYER && pVictim->GetTypeId() == TYPEID_PLAYER)
        hitChance = 95.0f + skillDiff * 0.04f;  // PvP misschance base is 5.00%
    else if (pVictim->GetTypeId() == TYPEID_PLAYER)
        hitChance = 95.0f + skillDiff * 0.04f;
    else if (skillDiff < -10)
        hitChance = 95.0f + skillDiff * 0.2f;        // 5% base chance with an increased scaling from skilldiff
    else
        hitChance = 95.0f + skillDiff * 0.1f;
```

The changes should obviously also apply to the normal attack formula here:
https://github.com/LightsHope/server/blob/development/src/game/Objects/Unit.cpp#L3159

This formula would result in the following table of miss chances vs a lvl 63 mob: https://gyazo.com/7c3890eacd55e11e868d8ccc7e96be23

This formula is also supported by the wowwiki entry for miss chance from 2006: http://wowwiki.wikia.com/wiki/Miss?oldid=347939

Reasoning for this new formula:

1. The base miss chance would be a static 5%, much like the other attack table components.
2. With 315 wepskill vs a lvl 63 opponent the miss chance would be 5%, not 5.6%. Makes more sense, because the extra 0.6% comes from 15 defense skill, but it should be negated by the extra 15 weapon skill.
3. Mathemathically it makes more sense for the miss chance at 300 weapon skill vs lvl 63 opponent to be 8% instead of 8.6%, since the 0.04% scaling would logically be increased to a different number, instead of the base miss chance being changed.

**TL:DR**
Base miss chance should be 5% instead of 5.6%

Miss chances vs lvl 63

- 300 weapon skill, 8% miss chance
- 305 weapon skill, 6% miss chance
- 310 weapon skill, 5.5% miss chance
- 315 weapon skill, 5% miss chance

 This would obviously still be increased by the +19% dual-wield miss chance penalty.

**Comment by GuybrushGit** [comment]
There is no reason to change this.

- The 300 skill miss chance was not 8%, there is footage of a rogue in AQ missing with 8%.
- That wowwiki entry was created by Baeza, who was notable for adding tons of information without any sort of evidence, much of it was later reverted.
- The base miss chance changing depending on skill difference was an actual mechanic, documented on elitistjerks. Our current miss formula already has unblizzlike changes to adjust for this, minimizing the effect of this "skill bump", see https://github.com/elysium-project/server/pull/759

**Comment by ohganot** [comment]
Where can I view the elitistjerks source?

This still doesn't change that fact that base miss chance shouldnt be 5.6%, you should have 5% chance to miss vs a lvl 63 when you have 315 weapon skill.

**Comment by GuybrushGit** [comment]
It used to be 5%, until that custom change to adjust for the bump I linked.
Here's one of the EJ threads where they researched all this: http://web.archive.org/web/20090715100718/http://elitistjerks.com/f31/t11885-rogue_warrior_weapon_skill_adjustment_discussion/p20/#post469576

**Comment by ohganot** [comment]
The post on elitistjerks is from TBC, so it isn't a reliable source.

TBC saw multiple changes to the attack table, including the addition of expertise rating, which also changed base dodge to 6.5% and parry chance even higher. I wouldn't be surprised if they also changed the miss chance system.

**Comment by GuybrushGit** [comment]
It is the only source. There was no in depth research on hit formulas during vanilla. 
There were however a few remarks which fit in with the TBC formula, such as the cap being above 8 and under or equal to 9.
Between the TBC formula and something completely made up the choice is obvious.

**Comment by ohganot** [comment]
The current formula is also made up.

But I think we can conclude that the scaling from 305 to 315 weapon skill should be 0,1% miss chance per point with a base 5% miss chance. That part of the formula is the same in the TBC formula and in the Baeza formula.

I made a list of currently known systems and also added some interesting combinations of those.
https://gyazo.com/c24d7f8cf37c5f506e76224322fc13c0

**Comment by king-tuts** [comment]
i don't understand how Cephel's pull request gets merged without providing any evidence, yet you are shooting this guy down for making a valid argument about base miss chance as he provides evidence to support it

**Comment by ohganot** [comment]
In any case, the previous change to the miss chance formula should be reverted, since it has no sources and causes a anomaly with the 5.6% miss chance at 315 weapon skill.

The arguments for the bump affecting DPS so much that it would have been noticed is also false. The main benefit from weapon skill comes from increased damage on glancing blows, not the decreased miss chance. 
Furthermore, it is incorrect to claim that half the weapon skill items would offer "zero benefit" due to hit chance only being able to be obtained in 1% intervals. For example, if you have 6% hit and 5.6% miss chance, you would gain total 5.6% reduction to miss chance, not 5% like Cephel claims.

**Comment by king-tuts** [comment]
Also, under the current formula, the base miss chance for 300 weapon skill against 300 defense is also 5.6% and this is definitely incorrect.

**Comment by GuybrushGit** [comment]
If you want to petition the team to revert the custom changes by Cephei then sure, you have my full support. But implementing one made up formula just because the current one is made up as well is not the right way to go about it.

**Comment by ohganot** [comment]
The point is, every formula is gonna be made up, because we don't have a vanilla source for it. What we do know, is that the base miss chance is supposed to be 5%, so we should have that correct atleast.

Now the question is if we want the miss chance to be 8%, 8.6% or 9% at 300 weapon skill. At 315 weapon skill it HAS to be 5% because thats the only thing we know for sure.

**Comment by fgsfdss** [comment]
@GuybrushGit in that post you linked the base miss chance is calculated as _5% + (Mob's Defense Skill - Your Weapon Skill) * .1%_, isn't it set at a flat 5.6% here? https://github.com/LightsHope/server/blob/development/src/game/Objects/Unit.cpp#L3164
This would only be correct on units three levels higher without defense or weapon skill modifications.

That post also suggests that miss chance against a target of three levels higher at 300 weapon skill and 315 defense respectively should be 9% and not 8.6% as it is currently, right?

Further in the thread there is another post which claims to confirm these formulas:
>When Def Skill - Weapon Skill <= 10:
>5% + (Def Skill - Weapon Skill) * .1%
>When Def Skill - Weapon Skill > 10:
>7% + (Def Skill - Weapon Skill - 10) * .4%

http://web.archive.org/web/20090711164229/http://elitistjerks.com/f31/t11885-rogue_warrior_weapon_skill_adjustment_discussion/p22/#post472228

A calculator was made based on these numbers: http://web.archive.org/web/20090727074431/http://sp00n.pytalhost.com:80/wow/weaponskill.htm

Am I misunderstanding something here?


PS: On page 28 there's also some good information about dodge, parry and block chances.

**Comment by ohganot** [comment]
Once again, that source is from TBC, so I don't think we should take it as a trustworthy source.

Before TBC, we know that the base chance for dodge, block, parry and miss was 5%. Each point of +defense increased those by 0.04%, resulting in bosses having 5.6% chance to dodge, block and parry. However, miss chance worked differently and the chance to miss was from 8% to 9%.

I could only find sources of players saying that the hit cap was 8% in vanilla, so I think we shouldn't use the TBC value of 9%. 

In TBC, the chance of a boss dodging or blocking was increased from 5.6% to 6.5% and the chance of parrying increased from 5.6% to 9%. I think it's likely that the chance of missing was also increased from 8% to 9%.
Furthermore, in WotLK, the chance to miss was changed to 8% (again?).

I think Beaza's formula is accurate, because it is exactly the same as the TBC formula for weapon skill values from 305 to 315, but instead of being 9% chance to miss at 300 weapon skill, it has 8% chance to miss.

**Comment by GuybrushGit** [comment]
The hit cap in vanilla was not 8%.

Here's a screenshot of a rogue missing on Vem with 8%:
http://web.archive.org/web/20081111013822/http://sp00n.pytalhost.com/misc/backstabmiss.jpg
This came from an EJ 2006 thread which is not available anymore.

On this post another person claimed to miss white hits with 8%:
http://forums.elitistjerks.com/forums/topic/7565-comeplete-summary-of-hit-caps/?page=4#comment-204930

The TBC formula is definitely not made up, its the result of extensive testing some of which can be seen in that TBC thread: 
https://web.archive.org/web/20080108010959/http://elitistjerks.com:80/471121-post512.html

There is no indication that this formula changed from vanilla to TBC and since we don't have any valid research from vanilla, the TBC formula is the closest we'll get to blizzlike hit chances.

**Comment by cudderi** [comment]
How can one know, that the "+8% chance to hit" from the picture "backstabmiss.jpg" does not include a +3% hit from scope? 

Up until patch 1.10, "Biznicks 247x128 Accurascope" also affected melee attacks. (Fixed in AQ patch) [1] 
It is a common bug that addons "Titanpanel: Item Bonuses" and "BonusScanner", count the bonus from the scope to the total chance to hit. This can be replicated on Anathema. when you download said addons. 

+8 hit chance with 153 NR set can be hard to achieve without DFT/accuria, and/or BiS NR gear. However, the attackpower +94 seen on the picture can be achieved by Blessed qiraji pugio +18ap, Core hound tooth +20ap and DFT +56ap = 94ap. Is there any other way to achieve the 56 ap other than dft? I could not think of a way, since it cant be 2x rogue ZG enchants, since there is only +1% dodge, (and 2x +100hp alrdy).

So to reiterate my question, has anyone read the original forum post (now unavailable), and can confirm that Surprise (the rogue in the picture) did not use a hit scope?

Link to Backstabmiss.jpg:
http://web.archive.org/web/20081111013822/http://sp00n.pytalhost.com/misc/backstabmiss.jpg 
[1]Source http://wowwiki.wikia.com/wiki/Patch_1.10.0
